runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5872,34 +5872,25 @@ class INTERPRETER_CLASS
 				_currentThread->tempSlot = (UDATA)index;
 				rc = THROW_AIOB;
 			} else {
-				j9object_t value = NULL;
+				j9object_t value = VM_ValueTypeHelpers::loadFlattenableArrayElement(_currentThread, _objectAccessBarrier, _objectAllocate, arrayref, index, true);
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 				J9Class *arrayrefClass = J9OBJECT_CLAZZ(_currentThread, arrayref);
-				if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
-					j9object_t newObjectRef = _objectAllocate.inlineAllocateObject(_currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, false, false);
-
-					if (NULL == newObjectRef) {
-						buildGenericSpecialStackFrame(REGISTER_ARGS, 0);
-						pushObjectInSpecialFrame(REGISTER_ARGS, arrayref);
-						updateVMStruct(REGISTER_ARGS);
-						newObjectRef = _vm->memoryManagerFunctions->J9AllocateObject(_currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
-						VMStructHasBeenUpdated(REGISTER_ARGS);
-						arrayref = popObjectInSpecialFrame(REGISTER_ARGS);
-						restoreGenericSpecialStackFrame(REGISTER_ARGS);
-						if (J9_UNEXPECTED(NULL == newObjectRef)) {
-							rc = THROW_HEAP_OOM;
-							return rc;
-						}
-						arrayrefClass = VM_VMHelpers::currentClass(arrayrefClass);
+				if (NULL == value && J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
```

Comment:
Removing the `J9_IS_J9CLASS_FLATTENED` will cause a system dump. I believe the reason is `value = _objectAccessBarrier.inlineIndexableObjectReadObject(currentThread, receiverObject, index);` is being run in all cases in which the value is not flattened. In the case that it does return null it will always hit the HEAP_OOM error (which should occur only if the slow path is used and returns null, not the non flattened path returns null) This was the simplest way I could come up with to prevent that.

Reply:
The problem is that you have one path for value types and non-valuetypes. Basically you need a
```
#ifdef valuetype
if (J9_IS_ValueType(arrayrefClass)) {
//your code here
} else
#endif /* valuetype */
{

}